lib/deep_preloader.rb

```diff
@@ -3,129 +3,277 @@
+  class WorklistEntry
+    attr_reader :model, :association_reflection, :child_spec
+
+    def initialize(model, association_reflection, child_spec)
+      @model = model
+      @association_reflection = association_reflection
+      @child_spec = child_spec
     end
 
-    # some models may have the association already loaded
-    loaded_models, unloaded_models = models.partition { |m| m.association(assoc_name).loaded? }
+    def association_name
+      @association_reflection.name
+    end
 
-    loaded_children = loaded_models.map { |m| m.association(assoc_name).target }.compact
-    children_by_key = loaded_children.index_by { |c| c.read_attribute(child_key) }
+    def association_macro
+      @association_reflection.macro
+    end
 
-    # Load children necessary to resolve unloaded models
-    unloaded_keys = unloaded_models
-                    .map { |m| m.read_attribute(parent_key) }
-                    .uniq
-                    .compact
-                    .reject { |k| children_by_key.has_key?(k) }
+    def loaded?
+      model.association(association_name).loaded?
+    end
 
-    if unloaded_keys.present?
-      child_scope.where(child_key => unloaded_keys).each do |c|
-        children_by_key[c.read_attribute(child_key)] = c
-      end
+    def belongs_to?
+      association_macro == :belongs_to
     end
 
-    if child_preload_spec.present? && children_by_key.present?
-      child_preload_spec.preload(children_by_key.values)
+    def collection?
+      association_macro == :has_many
     end
 
-    unloaded_models.each do |m|
-      target = children_by_key[m.read_attribute(parent_key)]
-      association = m.association(association_reflection.name)
-      association.target = target
-      association.set_inverse_instance(target) if target
+    def key
+      model.read_attribute(parent_key_column)
     end
-  end
 
-  def self.preload_multiple_association(models, association_reflection, child_scope, child_preload_spec)
-    assoc_name = association_reflection.name
+    # Conceal the difference between singular and collection associations so
+    # that `load_children` can always `group_by` the key
+    def children
+      target = model.association(association_name).target
 
-    unless association_reflection.macro == :has_many
-      raise "Unsupported association type #{reflection_type}"
+      if collection?
+        target
+      elsif target
+        [target]
+      else
+        []
+      end
     end
 
-    parent_key = association_reflection.active_record_primary_key
-    child_key  = association_reflection.foreign_key
-
-    # some models may have the association already loaded
-    loaded_models, unloaded_models = models.partition { |m| m.association(assoc_name).loaded? }
-    loaded_children = loaded_models.flat_map { |m| m.association(assoc_name).target }
+    def children=(targets)
+      if collection?
+        target = targets
+      else
+        raise "too many children" if targets.size > 1
```

Comment:
Consider something more verbose, to avoid "invalid type transitions".
```ruby
if targets.size > 1
  raise ArgumentError.new("Attempting to assign #{targets.length} children to singular relation #{association_name}")
end
```